File: paleo_workbench/ui/pages/well_map_panel.py
"""数据页内嵌的井位地图折叠面板。

The standalone 井位地图 page was absorbed into the Data page: this panel
hosts :class:`ProjectWellMapPage` (unchanged) under the asset table, with
a checkable header that folds the map away when vertical space is needed.
The same points also persist into the project as a dedicated vector map
document — see ``paleo_workbench.project.well_location_map``.
"""

from __future__ import annotations

from pathlib import Path
from typing import Any

from PySide6.QtCore import Qt
from PySide6.QtGui import QIcon
from PySide6.QtWidgets import QFrame, QHBoxLayout, QLabel, QToolButton, QVBoxLayout

from paleo_workbench.ui.pages.project_well_map_page import ProjectWellMapPage

_ICONS_DIR = Path(__file__).parent.parent / "assets" / "icons"


def _icon(name: str) -> QIcon:
    path = _ICONS_DIR / name
    return QIcon(str(path)) if path.exists() else QIcon()


class WellMapPanel(QFrame):
    """Collapsible section: header strip + embedded well-location map."""

    def __init__(self, parent=None, *, map_page: ProjectWellMapPage | None = None):
        super().__init__(parent)
        self.setObjectName("WellMapPanel")

        layout = QVBoxLayout(self)
        layout.setContentsMargins(0, 0, 0, 0)
        layout.setSpacing(0)

        header = QFrame()
        header.setObjectName("WellMapPanelHeader")
        header_layout = QHBoxLayout(header)
        header_layout.setContentsMargins(8, 2, 8, 2)
        header_layout.setSpacing(6)

        self.toggle_button = QToolButton()
        self.toggle_button.setObjectName("WellMapPanelToggle")
        self.toggle_button.setText("井位地图")
        self.toggle_button.setIcon(_icon("map/panel-bottom.svg"))
        self.toggle_button.setToolButtonStyle(Qt.ToolButtonStyle.ToolButtonTextBesideIcon)
        self.toggle_button.setCheckable(True)
        self.toggle_button.toggled.connect(self._on_toggled)

        self.count_label = QLabel()
        self.count_label.setObjectName("WellMapPanelCount")

        header_layout.addWidget(self.toggle_button)
        header_layout.addStretch(1)
        header_layout.addWidget(self.count_label)

        self._header = header
        self.map_page = map_page or ProjectWellMapPage()

        layout.addWidget(header)
        layout.addWidget(self.map_page, 1)

        # setChecked(False) on an already-unchecked button emits nothing, so
        # sync the initial collapsed state explicitly.
        self.set_collapsed(True)
        self._on_toggled(self.toggle_button.isChecked())

    def _on_toggled(self, checked: bool) -> None:
        self.map_page.setVisible(checked)
        self.toggle_button.setArrowType(
            Qt.ArrowType.DownArrow if checked else Qt.ArrowType.RightArrow
        )

    def set_collapsed(self, collapsed: bool) -> None:
        self.toggle_button.setChecked(not collapsed)

    def is_collapsed(self) -> bool:
        return not self.toggle_button.isChecked()

    def set_header_visible(self, visible: bool) -> None:
        """Hide the fold header while the map is a page's main content."""
        self._header.setVisible(visible)

    def expand_and_focus(self, well_id: str) -> None:
        """Tree/table → map: unfold the panel and center on the well."""
        self.set_collapsed(False)
        self.map_page.focus_well(well_id)

    def refresh_domain(self, project: Any) -> None:
        """Push the current document into the map (signature-gated, cheap)."""
        try:
            self.map_page.refresh_domain(project)
        except Exception:
            pass
        wells = getattr(project, "wells", None) or []
        reference_count = sum(
            1 for well in wells if getattr(well, "spatial_scope", "workarea") == "reference"
        )
        workarea_count = len(wells) - reference_count
        if reference_count:
            self.count_label.setText(f"{workarea_count} 口测区井 · {reference_count} 口参考井")
        else:
            self.count_label.setText(f"{workarea_count} 口井" if workarea_count else "")

File: libs/ui_wellseis/src/qt/well_map_panel.cpp
#include <pwb/ui_wellseis/qt/well_map_panel.hpp>

#include <QHBoxLayout>
#include <QLabel>
#include <QToolButton>
#include <QVBoxLayout>

#include <pwb/ui_wellseis/page_state.hpp>
#include <pwb/ui_wellseis/qt/project_well_map_page.hpp>

namespace pwb::ui_wellseis::qt {

WellMapPanel::WellMapPanel(QWidget* parent, ProjectWellMapPage* map_page)
    : QFrame(parent), map_page_(map_page) {
    setObjectName(QStringLiteral("WellMapPanel"));
    auto* layout = new QVBoxLayout(this);
    layout->setContentsMargins(0, 0, 0, 0);
    layout->setSpacing(0);

    header_ = new QWidget(this);
    header_layout_ = new QHBoxLayout(header_);
    header_layout_->setContentsMargins(8, 4, 8, 4);
    header_layout_->setSpacing(8);

    toggle_button_ = new QToolButton(header_);
    toggle_button_->setText(QStringLiteral("井位地图"));
    toggle_button_->setCheckable(true);
    toggle_button_->setChecked(true);
    toggle_button_->setArrowType(Qt::DownArrow);
    connect(toggle_button_, &QToolButton::toggled, this,
            [this](bool checked) { set_collapsed(!checked); });
    header_layout_->addWidget(toggle_button_);

    count_label_ = new QLabel(header_);
    count_label_->setStyleSheet(QStringLiteral("color: #6b6f76;"));
    header_layout_->addWidget(count_label_);
    header_layout_->addStretch(1);
    layout->addWidget(header_);

    body_ = new QWidget(this);
    auto* body_layout = new QVBoxLayout(body_);
    body_layout->setContentsMargins(0, 0, 0, 0);
    if (map_page_ == nullptr) {
        map_page_ = new ProjectWellMapPage(body_);
    }
    body_layout->addWidget(map_page_);
    layout->addWidget(body_, 1);
}

WellMapPanel::~WellMapPanel() = default;

void WellMapPanel::set_collapsed(bool collapsed) {
    collapsed_ = collapsed;
    body_->setVisible(!collapsed);
    toggle_button_->setChecked(!collapsed);
    toggle_button_->setArrowType(collapsed ? Qt::RightArrow
                                           : Qt::DownArrow);
}

bool WellMapPanel::is_collapsed() const {
    return collapsed_;
}

void WellMapPanel::set_header_visible(bool visible) {
    header_->setVisible(visible);
}

void WellMapPanel::add_header_button(QToolButton* button) {
    header_layout_->insertWidget(header_layout_->count() - 1, button);
}

void WellMapPanel::expand_and_focus(const std::string& well_id) {
    set_collapsed(false);
    map_page_->focus_well(well_id);
}

void WellMapPanel::refresh_domain(const ProjectSlice& project) {
    map_page_->set_project(project);
    count_label_->setText(QString::fromStdString(
        well_map_counts_text(well_map_counts(project.wells))));
}

ProjectWellMapPage* WellMapPanel::map_page() const {
    return map_page_;
}

QString WellMapPanel::count_text() const {
    return count_label_->text();
}

}  // namespace pwb::ui_wellseis::qt
